src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -198,6 +248,303 @@ ErrorIfIllegalPartitioningInTenantSchema(Oid parentRelationId, Oid partitionRela
+ * fir different table types.
+ */
+static SchemaTables *
+TablesUnderSchema(Oid schemaId)
+{
+	List *relationIdListInSchema = GetRelationIdListInSchema(schemaId);
+
+	SchemaTables *schemaTables = palloc0(sizeof(SchemaTables));
+	schemaTables->schemaId = schemaId;
+
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, relationIdListInSchema)
+	{
+		CitusTableCacheEntry *tableEntry = LookupCitusTableCacheEntry(relationId);
+		if (tableEntry == NULL)
+		{
+			schemaTables->localRelationIdList = lappend_oid(
+				schemaTables->localRelationIdList,
+				relationId);
+		}
+		else if (IsCitusTableType(relationId, HASH_DISTRIBUTED))
+		{
+			schemaTables->hashDistributedRelationIdList = lappend_oid(
+				schemaTables->hashDistributedRelationIdList, relationId);
+		}
+		else if (IsCitusTableType(relationId, RANGE_DISTRIBUTED))
+		{
+			schemaTables->rangeDistributedRelationIdList = lappend_oid(
+				schemaTables->rangeDistributedRelationIdList, relationId);
+		}
+		else if (IsCitusTableType(relationId, APPEND_DISTRIBUTED))
+		{
+			schemaTables->appendDistributedRelationIdList = lappend_oid(
+				schemaTables->appendDistributedRelationIdList, relationId);
+		}
+		else if (IsCitusTableType(relationId, SINGLE_SHARD_DISTRIBUTED))
+		{
+			schemaTables->singleShardRelationIdList = lappend_oid(
+				schemaTables->singleShardRelationIdList, relationId);
+		}
+		else if (IsCitusTableType(relationId, REFERENCE_TABLE))
+		{
+			schemaTables->referenceRelationIdList = lappend_oid(
+				schemaTables->referenceRelationIdList, relationId);
+		}
+		else if (IsCitusTableType(relationId, CITUS_LOCAL_TABLE))
+		{
+			schemaTables->citusLocalRelationIdList = lappend_oid(
+				schemaTables->citusLocalRelationIdList, relationId);
+		}
+	}
+	return schemaTables;
+}
+
+
+/*
+ * SchemaHasCitusTableType determines if tables under schema has any Citus
+ * table type given in the list.
+ */
+static bool
+SchemaHasCitusTableType(SchemaTables *schemaTables, List *citusTableTypes)
+{
+	CitusTableType citusTableType = 0;
+	foreach_int(citusTableType, citusTableTypes)
+	{
+		switch (citusTableType)
+		{
+			case HASH_DISTRIBUTED:
+			{
+				if (schemaTables->hashDistributedRelationIdList)
+				{
+					return true;
+				}
+				break;
+			}
+
+			case RANGE_DISTRIBUTED:
+			{
+				if (schemaTables->rangeDistributedRelationIdList)
+				{
+					return true;
+				}
+				break;
+			}
+
+			case APPEND_DISTRIBUTED:
+			{
+				if (schemaTables->appendDistributedRelationIdList)
+				{
+					return true;
+				}
+				break;
+			}
+
+			case SINGLE_SHARD_DISTRIBUTED:
+			{
+				if (schemaTables->singleShardRelationIdList)
+				{
+					return true;
+				}
+				break;
+			}
+
+			case REFERENCE_TABLE:
+			{
+				if (schemaTables->referenceRelationIdList)
+				{
+					return true;
+				}
+				break;
+			}
+
+			case CITUS_LOCAL_TABLE:
+			{
+				if (schemaTables->citusLocalRelationIdList)
+				{
+					return true;
+				}
+				break;
+			}
+
+			default:
+			{
+				break;
```

Comment:
Can we throw an error here to make sure that the callers don't provide anything different than we expect?

"unexpected table type when verifying schema tables"